packages/liveblocks-client/src/shallow.test.ts

```diff
@@ -0,0 +1,150 @@
+import fc from "fast-check";
+
+import shallow from "./shallow";
+
+const scalar = () => fc.jsonValue({ maxDepth: 0 });
+
+const complex = () =>
+  fc.anything().filter((value) => value !== null && typeof value === "object");
+
+describe("shallow", () => {
+  it("scalar values", () => {
+    expect(shallow(0, 0)).toBe(true);
+    expect(shallow("", "")).toBe(true);
+    expect(shallow("hi", "hi")).toBe(true);
+    expect(shallow(false, false)).toBe(true);
+    expect(shallow(true, true)).toBe(true);
+
+    expect(shallow(0, 1)).toBe(false);
+    expect(shallow(false, true)).toBe(false);
+    expect(shallow(true, false)).toBe(false);
+
+    // Weird exceptions
+    expect(shallow(NaN, NaN)).toBe(true);
+    expect(shallow(-0, +0)).toBe(false);
+  });
+
+  it("scalar values wrapped in list", () => {
+    expect(shallow([0], [0])).toBe(true);
+    expect(shallow([""], [""])).toBe(true);
+    expect(shallow(["hi"], ["hi"])).toBe(true);
+    expect(shallow([false], [false])).toBe(true);
+    expect(shallow([true], [true])).toBe(true);
+
+    expect(shallow([0], [1])).toBe(false);
+    expect(shallow([false], [true])).toBe(false);
+    expect(shallow([true], [false])).toBe(false);
+
+    // Weird exceptions   ],[
+    expect(shallow([NaN], [NaN])).toBe(true);
+    expect(shallow([-0], [+0])).toBe(false);
+  });
+
+  it("scalar values wrapped in objs", () => {
+    expect(shallow({ k: 0 }, { k: 0 })).toBe(true);
+    expect(shallow({ k: "" }, { k: "" })).toBe(true);
+    expect(shallow({ k: "hi" }, { k: "hi" })).toBe(true);
+    expect(shallow({ k: false }, { k: false })).toBe(true);
+    expect(shallow({ k: true }, { k: true })).toBe(true);
+
+    expect(shallow({ k: 0 }, { k: 1 })).toBe(false);
+    expect(shallow({ k: false }, { k: true })).toBe(false);
+    expect(shallow({ k: true }, { k: false })).toBe(false);
+
+    // Weird exceptions
+    expect(shallow({ k: NaN }, { k: NaN })).toBe(true);
+    expect(shallow({ k: -0 }, { k: +0 })).toBe(false);
+  });
```

Comment:
This first part is pretty classic unit testing: example-based tests. These are still super useful for general readability. The same inputs are tested every time.